examples/pde.hpp

```diff
@@ -342,6 +342,141 @@ double HalfCoeffecient::Eval(mfem::ElementTransformation& T,
+
+    mfem::DenseMatrix EvalKappa(int i);
+    mfem::Vector ComputeShapeCenter(const mfem::Element& el);
+    mfem::Vector ComputeLocalWeight(int i);
+public:
+    LocalTPFA(const mfem::ParMesh& mesh, const mfem::SparseMatrix& vert_edge)
+        : Q_(NULL), VQ_(NULL), mesh_(mesh), vert_edge_(vert_edge) { }
+
+    /// @param q inverse of permeability \f$ kappa^{-1} \f$ in Darcy's law
+    LocalTPFA(const mfem::ParMesh& mesh, const mfem::SparseMatrix& vert_edge,
+              mfem::Coefficient& q)
+        : Q_(&q), VQ_(NULL), mesh_(mesh), vert_edge_(vert_edge) { }
+
+    LocalTPFA(const mfem::ParMesh& mesh, const mfem::SparseMatrix& vert_edge,
+              mfem::VectorCoefficient& q)
+        : Q_(NULL), VQ_(&q), mesh_(mesh), vert_edge_(vert_edge) { }
+
+    /// Compute local edge weights for the corresponding graph Laplacian
+    std::vector<mfem::Vector> ComputeLocalWeights()
+    {
+        std::vector<mfem::Vector> local_weights(mesh_.GetNE());
+        for (int i = 0; i < mesh_.GetNE(); ++i)
+        {
+            local_weights[i] = ComputeLocalWeight(i);
+        }
+        return local_weights;
+    }
+};
+
+mfem::DenseMatrix LocalTPFA::EvalKappa(int i)
+{
+    const int dim = mesh_.Dimension();
+    const mfem::Element& el = *(mesh_.GetElement(i));
+    auto trans = const_cast<mfem::ParMesh&>(mesh_).GetElementTransformation(i);
+    const auto& ip = mfem::IntRules.Get(el.GetType(), 1).IntPoint(0);
+
+    mfem::DenseMatrix kappa(dim);
+
+    // Note that Q is kappa^{-1}
+    if (VQ_)
+    {
+        mfem::Vector vq(dim);
+        VQ_->Eval(vq, *trans, ip);
+        for (int d = 0; d < dim; ++d)
+            kappa(d, d) = 1.0 / vq(d);
+    }
+    else if (Q_)
+    {
+        double sq = Q_->Eval(*trans, ip);
+        for (int d = 0; d < dim; ++d)
+            kappa(d, d) = 1.0 / sq;
+    }
+    else
+    {
+        for (int d = 0; d < dim; ++d)
+            kappa(d, d) = 1.0;
+    }
+
+    return kappa;
+}
+
+mfem::Vector LocalTPFA::ComputeShapeCenter(const mfem::Element& el)
+{
+    const int dim = mesh_.Dimension();
+    const int num_verts = el.GetNVertices();
+
+    mfem::Vector center(dim);
+    center = 0.0;
+
+    for (int v = 0; v < num_verts; ++v)
+    {
+        const double* vert_coord = mesh_.GetVertex(el.GetVertices()[v]);
+        for (int d = 0; d < dim; ++d)
+        {
+            center[d] += vert_coord[d];
+        }
+    }
+    return center /= num_verts;
+}
+
+mfem::Vector LocalTPFA::ComputeLocalWeight(int i)
+{
+    const int dim = mesh_.Dimension();
+    const int num_facets = vert_edge_.RowSize(i);
+    const mfem::DenseMatrix kappa = EvalKappa(i);
+    const mfem::Vector cell_center = ComputeShapeCenter(*(mesh_.GetElement(i)));
+
+    mfem::Array<int> edges;
+    GetTableRow(vert_edge_, i, edges);
+
+    mfem::ParMesh& non_const_mesh = const_cast<mfem::ParMesh&>(mesh_);
+    mfem::Vector normal(dim), c_vector(dim);
+    mfem::Vector local_weight(num_facets);
+
+    for (int e = 0; e < num_facets; ++e)
```

Comment:
Is there a reason you're using `facet` here? This is terminology invented by someone for Parelag, I think `face` is better in this method.

Reply:
My interpretation on `facet` is that a `facet` of a n-dimensional polytope is a (n-1)-dimensional object on part of boundary (as in [wiki](https://en.wikipedia.org/wiki/Face_(geometry)#Facet_or_(n-1)-face)). So it fits in both 2D and 3D: 2D is an edge, 3D is a face. 

face should be clear enough from the context though (I think the reader can understand it as an edge in 2D), so I will make the corresponding change.

Reply:
I learned something new today! I guess either way is fine.